### PR TITLE
Hotkey support for MMapper client

### PR DESCRIPTION
**Pull request by KasparMetsa**
## Summary

Adds comprehensive hotkey/macro support to MMapper's integrated MUD client, allowing users to bind keyboard shortcuts to commands. Includes CLI commands for management and support for command separators to send multiple commands with one keypress.

### Hotkey System
- Bind any supported key (F1-F12, Numpad, Number row, ACCENT, HYPHEN, EQUAL) to MUD commands
- Support for modifier combinations: CTRL, SHIFT, ALT, META
- Sensible defaults for numpad movement and door actions (open/close/pick)
- Automatic key normalization ensures consistent modifier order (CTRL+SHIFT+ALT+META)
- O(1) hotkey lookup using `std::unordered_map` with custom hash

### Performance Optimizations
- Commands stored as `std::string` internally to minimize QString↔std::string conversions on the hot path
- `getCommandQString()` convenience methods for Qt UI callers
- Optimized key event handling with `ShortcutOverride` interception

### CLI Commands
- `_hotkey set KEY command` - Set a hotkey
- `_hotkey remove KEY` - Remove a hotkey
- `_hotkey config` - List all configured hotkeys
- `_hotkey keys` - Show available key names and modifiers
- `_hotkey reset` - Reset to default configuration
- `_config edit` - Open all configuration (including hotkeys) in external editor

### Command Separator
- Multi-character command separator support (e.g., `;;` to send `kill orc;;get all corpse`)
- Configurable in preferences with validation

### Other Improvements
- PageUp/PageDown keys scroll the display (with preview split behavior)
- "Auto-start client" option (disabled in WASM where client always auto-starts)
- OpenGL: Switched to 2D texture arrays for improved rendering performance
- Fixed NORTH/SOUTH climb dotted wall orientation bug
- Comprehensive unit tests for HotkeyManager

### Documentation
- WebAssembly build guide (`docs/WEBASSEMBLY_BUILD.md`)
- WASM build script with corrected paths (`scripts/build-wasm.sh`)
- clang-format CI check documentation in `AGENTS.md`

## Supported Keys

Modifier keys (CTRL, ALT, META, SHIFT) can be combined with any key below:

| Category | Keys |
|----------|------|
| Function keys | F1-F12 |
| Numpad digits | NUMPAD0-9 |
| Numpad operators | NUMPAD_SLASH, NUMPAD_ASTERISK, NUMPAD_MINUS, NUMPAD_PLUS, NUMPAD_PERIOD |
| Numpad navigation | ARROWUP, ARROWDOWN, ARROWLEFT, ARROWRIGHT, PAGEUP, PAGEDOWN, HOME, END, INSERT, KEY5 |
| Number row | 0-9 (require modifiers to work as hotkeys) |
| Misc | ACCENT (grave/tilde), HYPHEN, EQUAL |

## Default Hotkeys

```ini
# Basic movement (numpad)
NUMPAD8 = n
NUMPAD4 = w
NUMPAD6 = e
NUMPAD2 = s
NUMPAD_MINUS = u
NUMPAD_PLUS = d

# Open exit (CTRL+numpad)
CTRL+NUMPAD8 = open exit n
CTRL+NUMPAD4 = open exit w
CTRL+NUMPAD6 = open exit e
CTRL+NUMPAD2 = open exit s

# Close exit (ALT+numpad)
ALT+NUMPAD8 = close exit n
ALT+NUMPAD4 = close exit w
ALT+NUMPAD6 = close exit e
ALT+NUMPAD2 = close exit s

# Pick exit (SHIFT+numpad)
SHIFT+NUMPAD8 = pick exit n
SHIFT+NUMPAD4 = pick exit w
SHIFT+NUMPAD6 = pick exit e
SHIFT+NUMPAD2 = pick exit s

# Other actions
NUMPAD7 = look
NUMPAD9 = flee
NUMPAD1 = ride
NUMPAD3 = stand
NUMPAD0 = bash
```

## Testing
- Tested on Debian, Windows, macOS native build (15.7.2) and WASM using Chrome
- Comprehensive test suite covering:
  - Key normalization and modifier ordering
  - Import/export functionality
  - Invalid key rejection
  - Mutation behaviors

## Summary by Sourcery

Add a configurable hotkey system to the integrated MUD client, including command parsing, configuration editing, and input handling updates, along with WebAssembly build tooling and tests.

New Features:
- Introduce HotkeyManager to define, normalize, persist, and resolve client keyboard hotkeys with support for modifiers and numpad/navigation keys.
- Add CLI support for managing hotkeys via a `_hotkey` command (set/remove/config/keys/reset) and a `config edit` command that opens a remote editor for client configuration and hotkeys.
- Enable multi-character command separators for splitting and sending multiple commands from both direct input and hotkey-triggered commands.
- Add an option to automatically start the integrated client on application launch and connect to the MUD.

Bug Fixes:
- Fix PageUp/PageDown handling in the client so they scroll the display instead of being treated only as keypad movement or ignored.

Enhancements:
- Refactor input handling to classify keys centrally, support hotkey lookup, and route PageUp/PageDown to scroll the display while preserving history navigation and terminal-style shortcuts.
- Extend help handling to support subcommand-sensitive help text for `config` and new `hotkey` commands.
- Improve CPack version parsing to fall back to MMAPPER_VERSION when the git-derived version is not parseable.

Build:
- Include HotkeyManager sources and new parser module in the main build and test targets.
- Add WebAssembly build and run scripts (build-wasm.sh, server.py) to streamline WASM builds with Qt and Emscripten.

Documentation:
- Add a WebAssembly build guide describing Emscripten, Qt WASM setup, and daily build/run workflows.
- Document clang-format CI usage in AGENTS (per PR description).

Tests:
- Add a dedicated HotkeyManager test suite covering normalization, import/export, validation, duplicates, persistence, and direct key lookup.
- Wire HotkeyManager into existing parser, configuration, and map-related unit test binaries so tests use the new configuration component.

### CODE REVIEW

**Comment by KasparMetsa** [comment]
Addressed the review feedback:

1. **Key name validation**: Added `isValidBaseKey()` in HotkeyManager - invalid keys like `F13` or `NUMPDA8` are now rejected (normalizeKeyString returns empty string).

2. **Scripts moved**: `build-wasm.sh` and `server.py` moved to `scripts/` directory; updated WEBASSEMBLY_BUILD.md paths.

3. **fixup() backslash handling**: Now strips backslashes to match validate() behavior.

4. **_config detection**: Tightened to require exact match or space after command.

5. **Test brittleness**: Removed exact count assertion, now uses `QVERIFY(!empty())`.

**Comment by KasparMetsa** [comment]
@sourcery-ai review

File: src/client/inputwidget.cpp [hunk 1]
```diff
@@ -247,6 +585,162 @@ bool InputWidget::tryHistory(const int key)
     return false;
 }
 
+// Process _hotkey commands and return true if handled
+static bool processHotkeyCommand(const QString &input, InputWidgetOutputs &outputs)
+{
+    if (!input.startsWith("_hotkey")) {
```
[the remark this answers is not in the document]
Comment (KasparMetsa) [reply src/client/inputwidget.cpp]: Fixed: `_config` now uses strict matching: `input == "_config" || input.startsWith("_config ")` at line 566.

File: scripts/build-wasm.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+set -e
+
+# Source Emscripten environment
+# IMPORTANT: Change this path to match your emsdk installation location
+source "$HOME/dev/emsdk/emsdk_env.sh"
+
+# Paths - automatically detect script location
+MMAPPER_SRC="$(cd "$(dirname "$0")" && pwd)"
+QT_WASM="$MMAPPER_SRC/6.5.3/wasm_multithread"
+QT_HOST="$MMAPPER_SRC/6.5.3/macos"
```
[the remark this answers is not in the document]
Comment (KasparMetsa) [reply scripts/build-wasm.sh]: Fixed: Changed to \`MMAPPER_SRC="$(cd "$(dirname "$0")/.." && pwd)"\` to correctly resolve project root.

File: tests/TestHotkeyManager.cpp [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    QCOMPARE(manager.getCommand("F1"), QString()); // Should be cleared
+
+    // Test another import clears and replaces
+    manager.importFromCliFormat("_hotkey F12 stand\n");
+    QCOMPARE(manager.getAllHotkeys().size(), 1);
+    QCOMPARE(manager.getCommand("F10"), QString()); // Should be cleared
+    QCOMPARE(manager.getCommand("F12"), QString("stand"));
+}
+
+void TestHotkeyManager::importEdgeCasesTest()
```
[the remark this answers is not in the document]
Comment (KasparMetsa) [reply tests/TestHotkeyManager.cpp]: Fixed: Added commentPreservationTest() that verifies comments and formatting survive import/export round trip with correct ordering.

File: src/preferences/clientconfigpage.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            QMessageBox::warning(this,
+                                 tr("Import Warning"),
+                                 tr("No recognized sections found in file.\n\n"
+                                    "Expected sections: [Hotkeys]"));
+        }
+    };
+
+    if constexpr (CURRENT_PLATFORM == PlatformEnum::Wasm) {
+        // Use browser's native file upload dialog
+        QFileDialog::getOpenFileContent(nameFilter, processImportedFile);
```
Comment (nschimme) [inline src/preferences/clientconfigpage.cpp]: likewise here

File: src/client/inputwidget.cpp [hunk 5]
```diff
@@ -247,6 +560,83 @@ bool InputWidget::tryHistory(const int key)
     return false;
 }
 
+// Process _config commands and return true if handled
+static bool processConfigCommand(const QString &input, InputWidgetOutputs &outputs)
+{
+    if (!(input == "_config" || input.startsWith("_config "))) {
```
[the remark this answers is not in the document]
Comment (KasparMetsa) [reply src/client/inputwidget.cpp]: Fixed

**Comment by nschimme** [comment]
I've refactored this PR in https://github.com/MUME/MMapper/pull/445 so lets continue working on it there
